content/blog/datafusion-custom-parquet-index.md

```diff
@@ -0,0 +1,232 @@
+
+**But Parquet itself is extensible**: it tolerates unknown bytes after data pages and arbitrary key/value pairs in its footer. We can exploit those hooks to **embed** a small, per‑file distinct‑value index directly in the file—no extra files, no format forks, and no compatibility breakage.
+
+In the rest of this post, we’ll:
+
+1. Walk through the simple binary layout for a distinct‑value list.
+2. Show how to write it inline after the normal Parquet pages.
+3. Record its offset in the footer’s metadata map.
+4. Extend DataFusion’s `TableProvider` to discover and use that index for file‑level pruning.
+5. Verify everything still works in DuckDB via `read_parquet()`.
+
+---
+
+## Background
+
+Several examples in the DataFusion repository illustrate the benefits of using external indexes for pruning:
+
+* [`parquet_index.rs`](https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/parquet_index.rs)
+* [`advanced_parquet_index.rs`](https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/advanced_parquet_index.rs)
+
+Those demos work by building separate index files (Bloom filters, maps of distinct values) and associating them with Parquet files. While effective, this approach:
+
+* **Increases operational complexity:** Two files per dataset to track.
+* **Risks synchronization issues:** Removing or renaming one file breaks the index.
+* **Reduces portability:** Harder to share or move Parquet data when the index is external.
+
+Meanwhile, critics of Parquet’s extensibility point to the lack of a *standard* way to embed auxiliary data (see Amudai). But in practice, Parquet tolerates unknown content gracefully:
+
+* **Arbitrary metadata:** Key/value pairs in the footer are opaque to readers.
+* **Unused regions:** Bytes after data pages (before the Thrift footer) are ignored by standard readers.
+
+We’ll exploit both to embed our index inline.
+
+---
+
+## Motivation
+
+When scanning Parquet files, DataFusion (like other engines) reads row group metadata and then data pages sequentially. If you filter on a highly selective predicate (e.g., `category = 'foo'`), you may still incur I/O for files or row groups containing no matches.
+
+Embedding a lightweight, per‑file distinct‑value index enables DataFusion to skip entire files that cannot satisfy the filter:
+
+* **Format‑preserving:** Unknown footer keys are ignored by other readers.
+* **Efficient lookup:** Checking a small in‑memory set of values is far cheaper than disk I/O.
+* **Compact storage:** Store only the offset and payload, not a full duplicate of the data.
+
+---
+
+## High‑Level Design
```

Comment:
I suggest we keep the post at a higher level here and  omit the details of the special index's structure  (and refer readers to the example instead). 

In my mind the main points we are trying to get across in the article are:
1. Parquet is extensible with custom indexes
2. You can use DataFusion to write and read them

The idea is that readers will want to put their own special indexes in Parquet rather than using the particular implementation we have in the example. So I think focusing on the things they would have to do and de-emphasizing what is specific to the distinct values index would help get this point across better

Reply:
Great suggestion @alamb !